RFC: Prelease Versions and Experimental Features for API (#115)
buildpacks/rfcs: text/0000-lifecycle-experimental-api-mode.md

# Summary
[summary]: #summary

This RFC proposes the addition of a third API mode `experimental` and extends [RFC#49](https://github.com/buildpacks/rfcs/blob/main/text/0049-multi-api-lifecycle-descriptor.md).

# Definitions
[definitions]: #definitions

experimental - a new API mode that is marked as experimental. Unlike supported and deprecated APIs, this may be for an unreleased API. The purpose is to gather feedback that will be used to improve the design before it's finalized. It will not carry the same promise of stability or compatibility.

# Motivation
[motivation]: #motivation

In order to have a smooth transition to 1.0 of the spec, there needs to be a way for the project to test out these ideas in the wild and gather feedback. Even beyond 1.0, this desire still stands for testing out concepts in an organized fashion where the feedback can be provided. This feedback will help us improve the design before we finalize it. The quality of feedback we get from users about their real life usage of experimental APIs will help us avoid problems we may miss when approving RFCs.

> /AP

Reviewer: ```suggestion

```

# What it is
[what-it-is]: #what-it-is

`experimental` is a new API mode for testing out unreleased API / spec changes on their path to being finalized. An experimental API can not be `deprecated` or `supported` as defined in [RFC#49](https://github.com/buildpacks/rfcs/blob/main/text/0049-multi-api-lifecycle-descriptor.md).

## Lifecycle Descriptor

This RFC proposes the addition of the `experimental` key in `lifecycle.toml` following the proposed schema:

```
[apis]
[apis.buildpack]
  experimental = ["0.6-rc1", "1.0-alpha1"]
[apis.platform]
  experimental = ["1.0-alpha1"]

[lifecycle]
  version = "0.10.0"
```
`experimental`:
contains an array of experimental API versions
the APIs will correspond to specs that may not be released and in branches that are still being worked on.

> the API version must be of the format: `<major>.<minor>-<alphanumeric>`, i.e. `1.0-alpha1`.

Reviewer: Are the experimental versions meant to be "linear" (like, only one line of experiments per version). Or is it more a forking thing. That is, could we have both `0.1-stackpacks` and `0.1-newstuff` co-exist with neither experimental version containing the features of the other. And then merge them both into `0.1`?

> the API version must be of the format: `<major>.<minor>-<alphanumeric>`, i.e. `1.0-alpha1`.

Author: It's intentionally open ended right now, but my hope is linear.

Since experimental versions are defined with a different format, they will NOT automatically get upgraded to a `supported` release when the experimental API has finalized.

## Lifecycle Labels

> This will extend the builder or lifecycle image with the above descriptor file will extend the label `io.buildpacks.lifecycle.apis` with the value:

Reviewer: ```suggestion
This will extend the the `io.buildpacks.lifecycle.apis` label on the builder or lifecycle image with the values contained in the above descriptor file. For example:
```

```json
{
  "buildpack": {
    "experimental": ["0.6-rc1", "1.0-alpha1"],
  },
  "platform": {
    "experimental": ["1.0-alpha1"],
  }
}
```

## API Types
### Experimental APIs
API versions aren't limited to those defined in a spec release, but can refer to upcoming spec work happening in a branch.

> New `CNB_PLATFORM_EXPERIMENTAL_MODE` and `CNB_BUILDPACK_EXPERIMENTAL` mode environment variables will control experimental mode with:

Reviewer 2: ```suggestion
New `CNB_PLATFORM_EXPERIMENTAL_MODE` and `CNB_BUILDPACK_EXPERIMENTAL_MODE` environment variables will control experimental mode with:
```

 - **If** `CNB_PLATFORM_EXPERIMENTAL_MODE=warn`, **Then** print a warning and continue
 - **If** `CNB_PLATFORM_EXPERIMENTAL_MODE=error`, **Then** fail
 - **If** `CNB_PLATFORM_EXPERIMENTAL_MODE=silent`, **Then** continue w/o warning

**When** the `api` field in a `buildpack.toml` file is set to an API version in the deprecated buildpack API range the lifecycle shall:
 - **If** `CNB_BUILDPACK_EXPERIMENTAL_MODE` is unset, **Then** print a warning and continue
 - **If** `CNB_BUILDPACK_EXPERIMENTAL_MODE=warn`, **Then** print a warning and continue
 - **If** `CNB_BUILDPACK_EXPERIMENTAL_MODE=error`, **Then** fail
 - **If** `CNB_BUILDPACK_EXPERIMENTAL_MODE=silent`, **Then** continue w/o warning

# How it Works
[how-it-works]: #how-it-works

The technical implementation will follow what's outlined in [RFC#49](https://github.com/buildpacks/rfcs/blob/main/text/0049-multi-api-lifecycle-descriptor.md).

# Drawbacks
[drawbacks]: #drawbacks
This will add more complexity/overhead that most users will not use.

# Alternatives
[alternatives]: #alternatives

## Do Nothing
If we choose to do nothing, we run the risk of releasing less tested API/spec releases.

### Not a Different Version Format

> The Experimental API version format could instead follow the same format as deprecated and supported: `<major>-<minor>`. By doing this, users could automatically get upgraded from an experimental to supported/deprecated API version which may be surprising to them.

Reviewer: @ekcasey is your question about this alternative, or did the alternative prompt you to ask this about the main proposal.

Where would we be removing the experimental version from?

> The Experimental API version format could instead follow the same format as deprecated and supported: `<major>-<minor>`. By doing this, users could automatically get upgraded from an experimental to supported/deprecated API version which may be surprising to them.

Reviewer 3: > @ekcasey is your question about this alternative, or did the alternative prompt you to ask this about the main proposal?

The latter.

> Where would we be removing the experimental version from?

The `experimental` section of the next lifecycle release. I assume we wouldn't want to support every experimental API forever especially if we finalize and release the feature in a supported API?

> The Experimental API version format could instead follow the same format as deprecated and supported: `<major>-<minor>`. By doing this, users could automatically get upgraded from an experimental to supported/deprecated API version which may be surprising to them.

Reviewer: So you're saying when `lifecycle` releases 1.0, we would remove Platform/Buildpack API support for 1.0-alpha1?

I guess I was thinking it would be more like when `lifecycle` releases 1.0, we would remove Platform/Buildpack API support for 0.9-alpha1. That is:
* for any `<major>.<minor>` we would also support `<major>.<minor>-<experimental>`. 
* when we drop a given `<major>.<minor>` we would also drop the corresponding `<major>.<minor>-<experimental>`

> The Experimental API version format could instead follow the same format as deprecated and supported: `<major>-<minor>`. By doing this, users could automatically get upgraded from an experimental to supported/deprecated API version which may be surprising to them.

Reviewer 3: I agree with your description of what would happen, but we still remove the experimental version eventually right? Do we deprecate the experimental version before we remove it?

I guess this drawback confuses me but it implies some things I am necessarily comfortable with.

In the original multi-api RFC the idea of "experimental support" was that the definition of a experimental API was subject to change. I think this is now implying that the definition of `1.0-alpha1` is never allow to change, instead we would need to add a `1.0-alpha2`? If that is true
1. Do we really want to do that? The lifecycle could end up supporting many more APIs concurrently, adding complexity
2. Do "experimental" APIs even make sense as a category of APIs? Wouldn't it make more sense then to say that `1.0-alpha1` is first "supported" and later "deprecated" in that case?

> The Experimental API version format could instead follow the same format as deprecated and supported: `<major>-<minor>`. By doing this, users could automatically get upgraded from an experimental to supported/deprecated API version which may be surprising to them.

Author: > Do "experimental" APIs even make sense as a category of APIs? Wouldn't it make more sense then to say that 1.0-alpha1 is first "supported" and later "deprecated" in that case?

Like I said to @jromero's questions above, I'm fine with experimental not being its own section, but I do believe unlike both of those it carries a different support guarantees/breaking policies. This just makes that clearer by making it separate.

> I think this is now implying that the definition of 1.0-alpha1 is never allow to change, instead we would need to add a 1.0-alpha2

My hope is that we could branch/tag a spec, but maybe that overhead isn't worth it.

> Do we really want to do that? The lifecycle could end up supporting many more APIs concurrently, adding complexity

The key is to have a good/fast way to sunset experimental modes IMO.

> ### Unstable Features

Reviewer 3: While there are things I like about this alternative I worry about combinatorial explosion with the API version. Can every unstable feature be used with every support API, are the interfaces the same in each case? etc. etc.

Author: I do think it limits what is possible for experimental. There are definitely "safer" cross cutting features and some that aren't. I just listed it as an alternative b/c it's something we could consider.

Similar to [RFC#91](https://github.com/buildpacks/rfcs/pull/91), instead of tying features to a specific release, features can be tied behind flags for testing. There's a potential this could be complicated to do if the features are coupled to other changes going on in the API/spec. The benefit is folks can try out individual features in a more explicit way.

# Prior Art
[prior-art]: #prior-art

## Rust Unstable Features
Rust has the concept of [Unstable Features](https://doc.rust-lang.org/beta/unstable-book/the-unstable-book.html) where they have to be explicitly enabled in code. They're all documented in one place.

## Heroku API Stability Attribute
The Heroku API marks the status of every API endpoint with the [`stability' attribute](https://devcenter.heroku.com/articles/platform-api-reference#stability).

# Unresolved Questions
[unresolved-questions]: #unresolved-questions
How does an experimental version map to the branch?

> Is there a better word than "experimental"?

Reviewer: it's common to use `X` in place of "experimental if we want it to be shorter. But I don't think `CNB_PLATFORM_X_MODE` would be obvious to most people.

Also could use "preview"

# What it is
[what-it-is]: #what-it-is

This proposes two changes to Lifecycle:
* Prerelease API versions which is used for testing API / spec changes.
* `experimental` is a new API mode for testing implementation of the API similar to the modes defined in [RFC#49](https://github.com/buildpacks/rfcs/blob/main/text/0049-multi-api-lifecycle-descriptor.md).

## Prerelease API

A prerelease API is a non-finalized API but is in a testable state. The the API version must be of the format: `<major>.<minor>-<alphanumeric>`, i.e. `1.0-alpha1`. Though these aren't final, there will be a tagged and release version of the appropriate spec for it. This way users can see what is supported when it's used. In most cases, these will be cut from the corresponding branch in `buildpacks/spec`, i.e. Buildpack API `0.5-alpha1` would be cut from the `buildpack/0.5` branch.

When Lifecycle supports a prerelease API, it can be treated like any other API version and be used in any mode: `experimental`, `supported`, `deprecated`.

> 

Author: @ekcasey asked a good question in the WG today about the support policy, i.e. if supporting `1.2` implies that `1.0-alpha1` that came before it. We can follow most implementations of semver range rules where alphanumeric versions fall outside of the normal operators.

> ## Experimental API Mode

Author: From the WG meeting today, @ekcasey was pushing to have this removed until it's needed.

This is a new mode for testing implementation of a API. The API can be a full release or a prerelease API.

> New `CNB_PLATFORM_EXPERIMENTAL_MODE` and `CNB_BUILDPACK_EXPERIMENTAL` mode environment variables will control experimental mode with:

Reviewer 3: Why two variables?

Author: copy / paste error :(

allowed values: `warn`, `error`, `silent`
default value: `warn`

> **When** the `CNB_PLATFROM_API` environment variable is set to an API version in the deprecated platform API, the lifecycle shall:

Reviewer 4: ```suggestion
**When** the `CNB_PLATFORM_API` environment variable is set to an API version in the experimental platform API, the lifecycle shall:
```

 - **If** `CNB_PLATFORM_EXPERIMENTAL_MODE` is unset, **Then** print a warning and continue
 - **If** `CNB_PLATFORM_EXPERIMENTAL_MODE=warn`, **Then** print a warning and continue
 - **If** `CNB_PLATFORM_EXPERIMENTAL_MODE=error`, **Then** fail
 - **If** `CNB_PLATFORM_EXPERIMENTAL_MODE=silent`, **Then** continue w/o warning

> **When** the `api` field in a `buildpack.toml` file is set to an API version in the deprecated buildpack API range the lifecycle shall:

Reviewer 4: ```suggestion
**When** the `api` field in a `buildpack.toml` file is set to an API version in the experimental buildpack API range the lifecycle shall:
```

    "deprecated": ["0.5", "0.6-rc1"]
  }
}
```

# How it Works
[how-it-works]: #how-it-works

The technical implementation will follow what's outlined in [RFC#49](https://github.com/buildpacks/rfcs/blob/main/text/0049-multi-api-lifecycle-descriptor.md).

# Drawbacks
[drawbacks]: #drawbacks
This will add more complexity/overhead that most users will not use.

# Alternatives
[alternatives]: #alternatives

## Do Nothing
If we choose to do nothing, we run the risk of releasing less tested API/spec releases.

### Not a Different Version Format
The Experimental API version format could instead follow the same format as deprecated and supported: `<major>-<minor>`. By doing this, users could automatically get upgraded from an experimental to supported/deprecated API version which may be surprising to them.

### Unstable Features
Similar to [RFC#91](https://github.com/buildpacks/rfcs/pull/91), instead of tying features to a specific release, features can be tied behind flags for testing. There's a potential this could be complicated to do if the features are coupled to other changes going on in the API/spec. The benefit is folks can try out individual features in a more explicit way.

> ### Experimental Section in the API

Author: From the WG today, @sclevine requested this be moved into the actual proposal.

Reviewer 3: If we pull in unstable features do we need something like `CNB_PLATFORM_UNSTABLE_MODE` with `error`, `warn` and `silent` options?

Reviewer 3: Assuming we go ahead with the unstable features, I think it is potentially confusing that "experimental" in `pack` is analogous to "unstable" in lifecycle, while "experimental" in lifecycle is an entirely different concept.

I would prefer to call the unstable features experimental features and leave the experimental API as documented in this RFC out for now. Having so many different mechanism for experimental support seems like it could be potentially confusing to end users imho.

Author: @ekcasey I moved experimental features up into the proposal and moved experimental "mode" out into alternatives. Do you still have naming concerns?

Reviewer 3: looks good!
